src/quirks.rs

```diff
@@ -154,8 +165,30 @@ pub fn set_hostname(url: &mut Url, new_hostname: &str) -> Result<(), ()> {
     if url.cannot_be_a_base() {
         return Err(());
     }
-    let result = Parser::parse_host(Input::new(new_hostname), SchemeType::from(url.scheme()));
+    // Host parsing rules are strict,
+    // We don't want to trim the input
```

Comment:
Where is the spec for all of this? Is this what the parser spec does when given the hostname state override or whatever?

Reply:
Here are the two tests that fail when a trimmed input is passed : 

```
running 713 tests
......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................thread '"sc://x/".host = " " ' panicked at 'assertion failed: `(left == right)`
  left: `"sc:///"`,
 right: `"sc://x/"`', tests/data.rs:190:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
.........thread '"sc://x/".host = "\u{0}" Non-special scheme' panicked at 'assertion failed: `(left == right)`
  left: `"sc:///"`,
 right: `"sc://x/"`', tests/data.rs:190:5
...........F..F...........................................................
failures:

failures:
    "sc://x/".host = " " 
    "sc://x/".host = "\u{0}" Non-special scheme

test result: FAILED. 711 passed; 2 failed; 0 ignored; 0 measured
```
Analyzing one of them Is what lead me to change it : 
```
        {
            "comment": "Non-special scheme",
            "href": "sc://x/",
            "new_value": "\u0000",
            "expected": {
                "href": "sc://x/",
                "host": "x",
                "hostname": "x"
            }
        },
```
I think I read the correct section, but misimplemented it: https://url.spec.whatwg.org/#hostname-state

I think the relevant rule here is this one:
> Otherwise, if state override is given, buffer is the empty string, and either url includes credentials or url’s port is non-null, validation error, return.

The problem is I don't think there's a way for me to figure out state_override is given :/